mmv1/products/dialogflowcx/Flow.yaml

```diff
@@ -27,17 +27,35 @@ timeouts: !ruby/object:Api::Timeouts
   update_minutes: 40
 custom_code: !ruby/object:Provider::Terraform::CustomCode
   custom_import: templates/terraform/custom_import/dialogflowcx_flow.go.erb
-  pre_create: templates/terraform/pre_create/dialogflow_set_location.go.erb
+  pre_create: templates/terraform/pre_create/dialogflowcx_set_location_skip_default_obj.go.erb
   pre_update: templates/terraform/pre_create/dialogflow_set_location.go.erb
-  pre_delete: templates/terraform/pre_create/dialogflow_set_location.go.erb
+  pre_delete: templates/terraform/pre_delete/dialogflowcx_set_location_skip_default_obj.go.erb
   pre_read: templates/terraform/pre_create/dialogflow_set_location.go.erb
+virtual_fields:
+  - !ruby/object:Api::Type::Boolean
+    name: is_default_start_flow
+    immutable: true
```

Comment:
Currently [the resulting Go code for this field](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-9336-old..auto-pr-9336#diff-f159518f4e42da538ba7ddfc076ea15fa504d5419cc8446409409a13cf514a77) marks it as `Optional: true` in the resource's schema, and I think we instead need the new fields to be `Computed: true` (and not optional). Currently a user would be able to attempt to set a value for this field, and [the generated documentation](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-9336-old..auto-pr-9336#diff-0ba1d33306da3edaef76037c801183721e19da59f04cc611428e09953f4b9257) describes the field as if it's user configurable.

I'm afraid that this might require more updates to `mmv1/provider/terraform/virtual_fields.rb` and the templates! Thank you for adding the ability for virtual fields to be more nuanced.

Also, if the field was made to be Computed in the resource's schema then then I think marking it as `immutable: true` in Magic Modules (and therefore `ForceNew: true` in the templated Go code) isn't necessary. Though I may not have thought though all scenarios affected by that - what do you think?

Reply:
To clarify my understanding: `Computed: true` means it's output-only, right? Like, it's not a user-settable field?

I think I understand where you're coming from, and I did a bad job explaining the context so I'll try in different words 😅.

The thing _I specifically want_ is for it to be user-settable. It's the intended way for the user to specify that this terraform resource is used to manage the GCP object that Dialogflow creates for you when you create an Agent.

**In the current world**, you create an `google_dialogflow_cx_agent` resource and you get an Agent GCP object; but you also get a Flow GCP object automatically (the Default Start Flow). It's a full GCP object (except you're not allowed to delete it, trying to errors), and you can set all the Flow properties on it, same as other Flow objects. There's just no easy way to manage it through terraform; you have to go through two `terraform apply` steps (or `terraform apply -target="the_agent_resource"` + `terraform import ...`, or some other complicated process). And _even if_ you go through that complicated process and import it into a `google_dialogflow_cx_flow` resource so you can manage it, if you need to delete stuff, removing the `google_dialogflow_cx_flow` resource errors.

**With this change** you as the user can create a `google_dialogflow_cx_agent` resource and a `google_dialogflow_cx_flow` that manages the Default Start Flow, without going through any complicated process. You just set `is_default_start_flow = true` and then run `terraform apply` and everything _just works_ (I think; that's my intent anyway)

(everything above also applies to intents, the `google_dialogflow_cx_intent` resource with the Default Welcome Intent and Default Negative Intent GCP objects)

Reply:
PS I'm not super thrilled with this API, it feels unsatisfying in a way I'm not really able to articulate. An alternative is to make the `name` property user-settable + optional + read-from-API, then switch behavior based on special values (`000...00` or `000...01`); but that _also_ feels weird 😐

Reply:
> The thing I specifically want is for it to be user-settable

Ah yes, sorry I think I got confused looking at the code 🤦 . I was thinking too much about the import scenario, where the name is supplied and allows the `is_default_x` field to be computed (i.e. output-only, you're correct in your understanding of Computed: true). But of course if you don't want to perform an explicit import action you can't supply the `name`, as it's output only, so supplying a `is_default_x` argument is required. Gotcha.

I think it feels weird because adopting resources that aren't created by Terraform (outside of using import actions) is hard to cater for while still allowing usual CRUD actions. Here's an issue that could help address that general problem: https://github.com/hashicorp/terraform/issues/19017

It might be that to address the original issue we create a separate resource for the default flows/intents, so the resources touched by the PR don't need to address both use cases. I'll ask for others' opinions in my team!

Reply:
I'm actually going to the office hours in a few minutes!

Reply:
After some discussion at office hours, I'll try writing a PR with the `name` property being settable and switch "Create" for "Update" based on that; the person I talked to thinks there's precedent for that, which makes me feel a lot better about it.

I'm converting this PR back into a "draft" for now, and I'll try to flag you for the next one.

Reply:
I think this conversation's run its course? No changes needed as a result -> resolving.